libs/common-services/src/main/java/uk/gov/di/ipv/core/library/retry/Retry.java

```diff
@@ -0,0 +1,36 @@
+package uk.gov.di.ipv.core.library.retry;
+
+import uk.gov.di.ipv.core.library.exceptions.RetryException;
+
+public class Retry {
+
+    private Retry() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    public static <T> T runTaskWithBackoff(
+            Sleeper sleeper, int maxAttempts, int waitInterval, RetryableTask<T> task)
+            throws RetryException, InterruptedException {
+
+        if (maxAttempts < 1) {
+            throw new IllegalArgumentException("max attempts must be greater than 0");
+        }
+        if (waitInterval < 1) {
+            throw new IllegalArgumentException("wait interval must be greater than 0");
+        }
+        var count = 0;
+        while (count < maxAttempts) {
+            var isLastAttempt = (count + 1) >= maxAttempts;
+            var res = task.run(isLastAttempt);
+            if (res.isPresent()) {
+                return res.get();
+            }
+            if (isLastAttempt) {
+                throw new RetryException("max attempts reached for task: " + maxAttempts);
```

Comment:
This feels a bit counter-intuitive as it duplicates the throw below, which is actually unreachable

I'd probably move the wait inside an `if` instead, and probably just convert this to a `for` loop:
```
for (var attempt = 1; attempt <= maxAttempts; attempt++) {
    var res = task.run(attempt == maxAttempts);
    if (res.isPresent()) {
        return res.get();
    }
    if (attempt < maxAttempts) {
        var backoff = (long) (waitInterval * Math.pow(2, attempt - 1));
        sleeper.sleep(backoff);
    }
}
```